### PR TITLE
[SYCL][SPIRV] Fix return type of relational builtins

### PR DESCRIPTION
**Pull request by hewj03**
E.g. islessgreater(double16, double16) should return long16.

### CODE REVIEW

**Comment by hewj03** [comment]
how to run LLVM_Test_Suite_CUDA locally?

**Review by bader — APPROVED** [review]
@hewj03, IIRC, the change you are removing was done intentionally to overcome some issues in LLVM optimizations. It looks like SYCL-CTS tests on relational functions pass, so it might signal that issues were fixed. Could you elaborate more on the problem you are trying to solve with that patch, please?

With regards to llvm-test-suite failure on CUDA, I think we need to update implementation of SPIR-V built-ins in libclc. Please, check if they are supported by AMD GPU and update as well. CI doesn't validate AMD GPU yet. To run the test suite locally, please, read [README file](https://github.com/intel/llvm-test-suite/blob/intel/SYCL/README.md).

I'm approving the patch to launch BuildBot validation only.

**Comment by hewj03** [comment]
@bader The removed code becomes unnecessary if return type is changed.
The problem I'm trying to solve is when spirv is not generated during compilation, frontend IR was
```
  %call3.i.i.i = tail call spir_func <16 x i8> @_Z21__spirv_LessOrGreaterDv16_dS_(<16 x double> %element1.sroa.0.0.copyload.i, <16 x double> %element2.sroa.0.0.copyload.i) #9
```
After "Translate SPIR-V builtins to OCL 2.0 builtins (spvtoocl20)" pass:
```
  %call3.i.i.i = call spir_func <16 x i8> @_Z13islessgreaterDv16_dS_(<16 x double> %element1.sroa.0.0.copyload.i, <16 x double> %element2.sroa.0.0.copyload.i) #6
```
The return type is not the same as OpenCL builtin islessgreater. I think it is better to fix like this patch since it follows the return value types as stated in sycl spec.


For spirv path, llvm-spirv translator previously translated
```
  %call3.i.i.i = tail call spir_func <16 x i8> @_Z21__spirv_LessOrGreaterDv16_dS_(<16 x double> %element1.sroa.0.0.copyload.i, <16 x double> %element2.sroa.0.0.copyload.i) #9
```
into
```
  %16 = call spir_func <16 x i64> @_Z13islessgreaterDv16_dS_(<16 x double> %element1.sroa.0.0.copyload.i, <16 x double> %element2.sroa.0.0.copyload.i) #0
```
This explains why previously opencl backend works.
Since https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/1220, llvm-spirv translate it into:
```
  %16 = fcmp one <16 x double> %element1.sroa.0.0.copyload.i, %element2.sroa.0.0.copyload.i
```
I don't have AMD GPU machine to test at the moment. Will update when I have.

File: sycl/include/CL/sycl/detail/generic_type_traits.hpp [hunk 1]
```diff
@@ -558,25 +547,8 @@ template <typename T>
...
-    typename ret_t::vector_t result(0);
-    for (size_t I = 0; I < N; ++I) {
-      result[I] = value[I];
-    }
-    return result;
-#else
-    return value;
-#endif
-  }
+  static ret_t apply(ret_t value) { return value; }
```
[the remark this answers is not in the document]
Comment (hewj03) [reply sycl/include/CL/sycl/detail/generic_type_traits.hpp]: removed

**Review by steffenlarsen — DISMISSED** [review]
LGTM. @hewj03 would you mind opening an issue for the CUDA (and potentially HIP) backend failures this introduces once it is merged?

**Comment by smanna12** [comment]
Can you add tests?

**Comment by hewj03** [comment]
> Can you please add test?

done

**Comment by hewj03** [comment]
> LGTM. @hewj03 would you mind opening an issue for the CUDA (and potentially HIP) backend failures this introduces once it is merged?

done. I can't find previous fail log in this page, so the description is not clear.

File: sycl/test/check_device_code/builtin-relational.cpp [hunk 2]
```diff
@@ -0,0 +1,100 @@
+// RUN: %clangxx -I %sycl_include -S -emit-llvm -fsycl-device-only -fsycl-unnamed-lambda -Xclang -disable-llvm-passes %s -o - | FileCheck %s
+
```
[the remark this answers is not in the document]
Comment (hewj03) [reply sycl/test/check_device_code/builtin-relational.cpp]: done

**Review by steffenlarsen — APPROVED** [review]

**Review by smanna12 — APPROVED** [review]

File: sycl/test/check_device_code/builtin-relational.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (smanna12) [reply sycl/test/check_device_code/builtin-relational.cpp]: Thank you.

**Comment by Naghasan** [comment]
These builtins should return bool or vector of bool

e.g. https://www.khronos.org/registry/SPIR-V/specs/unified1/SPIRV.html#_a_id_relational_and_logical_a_relational_and_logical_instructions

```
OpIsNan

Result is true if x is an IEEE NaN, otherwise result is false.

Result Type must be a scalar or vector of Boolean type.

x must be a scalar or vector of floating-point type. It must have the same number of components as Result Type.

Results are computed per component.
```

At the time this file was written, LLVM banned vectors of i1 (bool), so it was mapped to i8.

**Comment by hewj03** [comment]
thank you @Naghasan, that makes sense, I'll close this PR.
@bader the loop in RelConverter::apply is generating inefficient IR, will sycl spec also change relational builtin type to bool?

**Comment by bader** [comment]
> will sycl spec also change relational builtin type to bool?

I'm not aware of such plans.

**Pull request closed without merging**